Matter_Engine/Query/Density.cs

```diff
@@ -0,0 +1,160 @@
+        {
+            Output<double, string> result = IDensityWithReport(materialProp, temprature, humidity);
+            Reflection.Compute.RecordWarning(result.Item2);
+            return result.Item1;
+        }
+
+        /***************************************************/
+
+        private static Output<double, string> IDensityWithReport(this IMaterialProperties materialProp, double temprature, double humidity)
+        {
+            object density = null;
+            Output<double, string> result = new Output<double, string>();
+
+            density = Reflection.Compute.RunExtensionMethod(materialProp, "Density", new object[] { temprature, humidity });
+            if (density != null)
+            {
+                return new Output<double, string>()
+                {
+                    Item1 = System.Convert.ToDouble(density),
+                    Item2 = ""
+                };
+            }
+            // Does not work without more sofisticated use of reflection to check input param names
+
+            //density = Reflection.Compute.RunExtensionMethod(materialProp, "Density", new object[] { temprature });
+            //if (density != null)
+            //{
+            //    return new Output<double, string>()
+            //    {
+            //        Item1 = System.Convert.ToDouble(density),
+            //        Item2 = UnUsedVaribles(materialProp.Name, "humidity")
+            //    };
+            //}
+
+            //density = Reflection.Compute.RunExtensionMethod(materialProp, "Density", new object[] { humidity });
+            //if (density != null)
+            //{
+            //    return new Output<double, string>()
+            //    {
+            //        Item1 = System.Convert.ToDouble(density),
+            //        Item2 = UnUsedVaribles(materialProp.Name, "temprature")
+            //    };
+            //}
+
+            density = Reflection.Compute.RunExtensionMethod(materialProp, "Density");
+            if (density != null)
+            {
+                return new Output<double, string>()
+                {
+                    Item1 = System.Convert.ToDouble(density),
+                    Item2 = UnUsedVaribles(materialProp.Name, "temprature and humidity")
+                };
+            }
+
+            density = Reflection.Query.PropertyValue(materialProp, "Density");
+            if (density != null)
+            {
+                return new Output<double, string>()
+                {
+                    Item1 = System.Convert.ToDouble(density),
+                    Item2 = "The value of the density for the MaterialFragment: " + materialProp.Name + "was acquired through its properties"
+                };
+            }
+
+            return new Output<double, string>()
+            {
+                Item1 = double.NaN,
+                Item2 = "The density could not be acquired from the MaterialFragment " + materialProp.Name
+            };
+        }
```

Comment:
### Proposed kind of method concept
Materials are complex and BHoM has no centralised stance on their definition.

The description of the materials will by necessity be a simplification, which is fine as long as the assumptions made are clear. Having this high level engine which grabs whichever material information it can from the fragments provided runs the risk of distancing the users from the assumptions. 

The idea here is that the methods would provide continuous documentation on the assumptions made and collect it as a single formatted document/message. I believe outputting notes would quickly fill up the available amount of notes possible to display, keep the information locked in the canvas and make it easily overlooked, especially in hindsight.

This method currently works by attempting to get the information from a method with the most amount of inputs and then tries those with less inputs and finally if there's a property.